File: packages/shared/src/domain/entities/PartnerProposal.ts
import type { PartnerRecruitmentStatus } from './PartnerRecruitment'

/** 서버가 응답·철회·경과 시간·모집 상태로 계산한 제안 상태입니다. 화면은 다시 계산하지 않습니다. */
export type PartnerProposalStatus = 'PENDING' | 'ACCEPTED' | 'DECLINED' | 'WITHDRAWN' | 'EXPIRED'

export const partnerProposalStatusLabels: Record<PartnerProposalStatus, string> = {
  PENDING: '응답 대기',
  ACCEPTED: '수락',
  DECLINED: '거절',
  WITHDRAWN: '철회',
  EXPIRED: '만료',
}

/** 상태 칩의 색 역할입니다. 화면들이 같은 매핑을 씁니다. */
export const partnerProposalStatusTones: Record<PartnerProposalStatus, 'ok' | 'warn' | 'muted' | 'info'> = {
  PENDING: 'info',
  ACCEPTED: 'ok',
  DECLINED: 'muted',
  WITHDRAWN: 'muted',
  EXPIRED: 'warn',
}

export const proposalMessageMaxLength = 500

/** 받은 제안함과 보낸 제안함입니다. */
export type PartnerProposalBox = 'received' | 'sent'

/** 상대 기업입니다. 기본정보는 프로필 공유를 켰거나 수락됐을 때, 담당자 연락처는 수락됐을 때만 옵니다. */
export type PartnerProposalCounterpart = {
  companyName: string
  isEmailVerified: boolean
  isBusinessVerified: boolean
  profile: { region: string; industry: string; foundedYear: number; homepageUrl: string | null } | null
  contact: { email: string; businessNumber: string } | null
}

export type PartnerProposal = {
  id: number
  status: PartnerProposalStatus
  message: string
  shareProfile: boolean
  /** 내가 보낸 제안이면 true, 내 모집글로 받은 제안이면 false입니다. */
  isSent: boolean
  recruitment: {
    id: number
    title: string
    status: PartnerRecruitmentStatus
    /** YYYY-MM-DD */
    recruitmentDeadline: string
  }
  counterpart: PartnerProposalCounterpart
  createdAt: string
  expiresAt: string
  respondedAt: string | null
}

export type PartnerProposalInput = {
  message: string
  shareProfile: boolean
}

/** 제안함 한 상자입니다. 대기 건수는 배지에 씁니다. */
export type PartnerProposalBoxPage = {
  box: PartnerProposalBox
  proposals: PartnerProposal[]
  pendingCount: number
}

/** 모집글 상세에 실리는 내 제안 요약입니다. 제안한 적이 없으면 null입니다. */
export type MyPartnerProposal = {
  id: number
  status: PartnerProposalStatus
}

File: backend/core-service/src/main/kotlin/ai/govbiz/core/partner/domain/PartnerProposal.kt
package ai.govbiz.core.partner.domain

import java.time.Duration
import java.time.LocalDate
import java.time.LocalDateTime

/** 저장하지 않고 응답·철회·경과 시간·모집 상태로 조회 시점에 계산하는 제안 상태입니다. */
enum class PartnerProposalStatus {
    PENDING,
    ACCEPTED,
    DECLINED,
    WITHDRAWN,
    EXPIRED,
}

/** 제안함 종류입니다. 받은 제안은 내 모집글로 온 것, 보낸 제안은 내가 제안자인 것입니다. */
enum class PartnerProposalBox {
    RECEIVED,
    SENT,
}

/** 조회 시각 기준으로 상태와 연락처 공개 여부를 계산해 둔 제안입니다. Service가 만들고 응답은 이 값을 그대로 씁니다. */
data class PartnerProposalView(
    val proposal: PartnerProposal,
    val status: PartnerProposalStatus,
    val recruitmentStatus: PartnerRecruitmentStatus,
    val revealsContacts: Boolean,
)

/** 모집글 작성자가 내린 결정입니다. 저장되는 두 값만 가집니다. */
enum class PartnerProposalDecision {
    ACCEPTED,
    DECLINED,
}

/** 제안자가 입력하는 내용입니다. */
data class PartnerProposalInput(
    val message: String,
    val shareProfile: Boolean,
) {
    init {
        require(message.isNotBlank() && message == message.trim() && message.length <= MAX_MESSAGE_LENGTH) {
            "message must be a trimmed 1~$MAX_MESSAGE_LENGTH character text"
        }
    }

    companion object {
        const val MAX_MESSAGE_LENGTH = 500
    }
}

/** 저장 직전의 새 제안입니다. 모집글·기업은 Service가 확인한 식별자를 넘깁니다. */
data class NewPartnerProposal(
    val recruitmentId: Long,
    val proposerAccountId: Long,
    val proposerCompanyId: Long,
    val content: PartnerProposalInput,
)

/** 제안의 한쪽 당사자입니다. 이메일은 수락된 뒤에만 상대에게 보여 줍니다. */
data class PartnerProposalParty(
    val accountId: Long,
    val email: String,
    val companyName: String,
    val businessNumber: String,
    val region: String,
    val industry: String,
    val foundedYear: Int,
    val homepageUrl: String?,
    val isEmailVerified: Boolean,
)

/** 제안이 묶인 모집글의 요약입니다. 모집이 끝나면 대기 중 제안도 만료로 봅니다. */
data class PartnerProposalRecruitment(
    val id: Long,
    val title: String,
    val recruitmentDeadline: LocalDate,
    val closedAt: LocalDateTime?,
    val program: PartnerRecruitmentProgram,
) {
    fun status(today: LocalDate): PartnerRecruitmentStatus =
        resolveRecruitmentStatus(closedAt, recruitmentDeadline, program, today)
}

/** 저장된 제안입니다. 담당자 연락처 공개 여부는 상태로만 정합니다. */
data class PartnerProposal(
    val id: Long,
    val recruitment: PartnerProposalRecruitment,
    val proposer: PartnerProposalParty,
    val owner: PartnerProposalParty,
    val content: PartnerProposalInput,
    val decision: PartnerProposalDecision?,
    val respondedAt: LocalDateTime?,
    val withdrawnAt: LocalDateTime?,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime,
) {
    /** 응답이 없으면 이 시각에 만료됩니다. */
    val expiresAt: LocalDateTime
        get() = createdAt.plus(RESPONSE_WINDOW)

    fun status(now: LocalDateTime): PartnerProposalStatus =
        when {
            withdrawnAt != null -> PartnerProposalStatus.WITHDRAWN
            decision == PartnerProposalDecision.ACCEPTED -> PartnerProposalStatus.ACCEPTED
            decision == PartnerProposalDecision.DECLINED -> PartnerProposalStatus.DECLINED
            now.isAfter(expiresAt) -> PartnerProposalStatus.EXPIRED
            recruitment.status(now.toLocalDate()) == PartnerRecruitmentStatus.CLOSED -> PartnerProposalStatus.EXPIRED
            else -> PartnerProposalStatus.PENDING
        }

    fun isProposedBy(accountId: Long): Boolean = proposer.accountId == accountId

    fun isOwnedBy(accountId: Long): Boolean = owner.accountId == accountId

    /** 수락된 제안만 양쪽 담당자 이메일과 기업 기본정보를 서로에게 공개합니다. */
    fun revealsContacts(now: LocalDateTime): Boolean = status(now) == PartnerProposalStatus.ACCEPTED

    companion object {
        val RESPONSE_WINDOW: Duration = Duration.ofDays(7)
    }
}
